webapp/src/ts/services/search-telemetry.service.ts

```diff
@@ -0,0 +1,73 @@
+import { Injectable } from '@angular/core';
+
+import { TelemetryService } from '@mm-services/telemetry.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class SearchTelemetryService {
+  private readonly CONTACT_SKIPPED_PROPERTIES = ['_id', '_rev', 'type', 'refid', 'geolocation'];
+  private readonly REPORT_SKIPPED_PROPERTIES = ['_id', '_rev', 'type', 'refid', 'content'];
+
+  constructor(
+    private readonly telemetryService: TelemetryService,
+  ) {}
+
+  public async recordContactSearch(contactDoc: Record<string, any>, search: string) {
+    const matchingProperties = await this.findMatchingProperties(contactDoc, search, this.CONTACT_SKIPPED_PROPERTIES);
+
+    await Promise.all(
+      matchingProperties.map(key => this.telemetryService.record(`search_match:contacts_by_freetext:${key}`)),
+    );
+  }
+
+  public async recordContactByTypeSearch(contactDoc: Record<string, any>, search: string) {
+    const matchingProperties = await this.findMatchingProperties(contactDoc, search, this.CONTACT_SKIPPED_PROPERTIES);
+
+    await Promise.all(
+      matchingProperties.map(key => this.telemetryService.record(`search_match:contacts_by_type_freetext:${key}`)),
+    );
+  }
+
+  public async recordReportSearch(reportDoc: Record<string, any>, search: string) {
+    const [matches, fieldsMatches] = await Promise.all([
+      this.findMatchingProperties(reportDoc, search, this.REPORT_SKIPPED_PROPERTIES),
+      this.findMatchingProperties(reportDoc.fields, search, this.REPORT_SKIPPED_PROPERTIES, 'fields'),
+    ]);
+    const matchingProperties = Array.from(new Set([...matches, ...fieldsMatches]));
+
+    await Promise.all(
+      matchingProperties.map(key => this.telemetryService.record(`search_match:reports_by_freetext:${key}`)),
+    );
+  }
+
+  private async findMatchingProperties(
+    doc: Record<string, any>,
+    search: string,
+    skip: string[],
+    basePropertyPath = '',
+  ) {
+    const colonSearch = search.split(':');
+    if (colonSearch.length > 1) {
+      return [`${colonSearch[0]}:$value`];
```

Comment:
After reading your proposal I was a bit worried that we'd lose out on the `key:value` searches but we kind of reached the conclusion that it's not used by anyone so it makes sense to not handle these queries. It would have been nice to have the data to back up that claim but risking to leak PII to get this data is not worth it.
So instead of continuing the search with `colonSearch[1]`, I think we should remove the `colonSearch`-related code from line 50 to 53, what do you think @jkuester?

Reply:
Coming back to this, I think your second approach is better. So removing line 50-53 and at the end of the function, re-run it with `colonSearch[1]` only if `matchingProperties` is empty.